# Evaluating RAG Systems

## 1. Introduction

In this task, we will explore how to evaluate the performance and quality of two different RAG implementations. The domain we’ll be dealing with is video transcripts—our models will answer questions based on the content from these transcripts.

All video transcriptions are from https://www.youtube.com/channel/UCmy-241n-ZLO4c5BA5W516g channel, what is official for [EPAM DIAL AI](https://epam-rail.com/)


## Objective:
By the end of this exercise, you will:
1.	Understand how to establish meaningful metrics to evaluate retrieval and generation quality.
2.	Create a test dataset from the provided transcripts.
3.	Evaluate two different RAG approaches to see which performs best.
4.	Document your findings and propose improvements.

This exercise should take approximately 2–4 hours to complete. You’ll be able to apply these methods directly to real-world RAG tasks in your day-to-day projects.


In [3]:
!pip install -U langchain-core pandas langchain-openai

Looking in indexes: https://pypi.org/simple, https://auto_epm-ease_git_integration%40epam.com:****@eu.artifactory.epam.com/artifactory/api/pypi/EPM-ESP-pypi/simple
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.5/79.5 kB 5.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 109.4/109.4 kB 9.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 548.1/548.1 kB 33.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 101.6 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.6/99.6 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 399.6/399.6 kB 28.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 56.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 472.3/472.3 kB 27.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 69.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [5]:
!pip show langchain-core

Name: langchain-core
Version: 1.4.0
Summary: Building applications with LLMs through composability
Home-page: 
Author: 
Author-email: 
License: MIT
Location: /opt/conda/lib/python3.11/site-packages
Requires: jsonpatch, langchain-protocol, langsmith, packaging, pydantic, pyyaml, tenacity, typing-extensions, uuid-utils
Required-by: deepeval, langchain, langchain-community, langchain-openai, langchain-text-splitters, ragas


In [7]:
!pip install -U langchain

Looking in indexes: https://pypi.org/simple, https://auto_epm-ease_git_integration%40epam.com:****@eu.artifactory.epam.com/artifactory/api/pypi/EPM-ESP-pypi/simple
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.0/42.0 kB 3.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.3/114.3 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.3/235.3 kB 13.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.2/56.2 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.0/41.0 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.0/134.0 kB 12.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.4/212.4 kB 15.5 MB/s eta 0:00:00
  Attempting uninstall: orjson
    Found existing installation: orjson 3.10.7
    Uninstalling orjson-3.10.7:
      Successfully uninstalled orjson-3.10.7
  Attempting uninstall: langchain
    Found 

In [10]:
import langchain_core.vectorstores as vs
print(dir(vs))

['ABC', 'Any', 'BaseRetriever', 'Callable', 'ClassVar', 'Collection', 'Dict', 'Embeddings', 'Field', 'Iterable', 'List', 'Optional', 'Sequence', 'TYPE_CHECKING', 'Tuple', 'Type', 'TypeVar', 'VST', 'VectorStore', 'VectorStoreRetriever', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'abstractmethod', 'annotations', 'logger', 'logging', 'math', 'root_validator', 'run_in_executor', 'warnings']


In [4]:
###
### NOTE: YOU DON'T NEED TO MODIRY THIS CELL
###      JUST RUN IT TO LOAD THE DATA AND CODE
###

import os
from collections import defaultdict
from typing import List

import pandas as pd
from langchain_core.documents import Document
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import RunnablePassthrough
from langchain_core.vectorstores import InMemoryVectorStore
from langchain_openai import AzureChatOpenAI, AzureOpenAIEmbeddings

# The key will be provided to you by leap platform automatically
API_KEY = os.environ["OPENAI_API_KEY"]
ENDPOINT = "https://ai-proxy.lab.epam.com"
EMBEDDINGS_MODEL = "text-embedding-3-small-1"
GENERATION_MODEL = "gpt-4o"
API_VERSION = "2023-07-01-preview"


def group_documents(documents: List[Document], n: int) -> List[Document]:
    """
    Groups documents (by 'title'), in ascending order by 'segment',
    chunking them every 'n' docs. Each output Document has:
      - concatenated page_content
      - updated metadata:
          * segment = the segment of the first doc in that group (or you can choose another naming)
          * start = start time of the first doc
          * end   = end time of the last doc
          * title/url = carried over from the group
    """
    # Separate documents by title
    docs_by_title = defaultdict(list)
    for doc in documents:
        docs_by_title[doc.metadata["title"]].append(doc)

    # Sort each title's documents by 'segment'
    for title in docs_by_title:
        docs_by_title[title].sort(key=lambda x: x.metadata["segment"])

    # Now chunk and group
    grouped_docs = []
    for title, docs in docs_by_title.items():
        for i in range(0, len(docs), n):
            chunk = docs[i : i + n]

            # Merge the page_content
            merged_content = " ".join(d.page_content for d in chunk)

            # Create new metadata
            new_metadata = {
                "segment": i // n,
                "start": chunk[0].metadata["start"],
                "end": chunk[-1].metadata["end"],
                "title": title,
                "url": chunk[0].metadata["url"],
                "segments_merged": [d.metadata["segment"] for d in chunk],
            }

            new_doc = Document(page_content=merged_content, metadata=new_metadata)
            grouped_docs.append(new_doc)

    return grouped_docs


def load_data(path):
    data = []
    for filename in os.listdir(path):
        if filename.endswith(".csv"):
            df = pd.read_csv(os.path.join(path, filename))
            for _, row in df.iterrows():
                document = Document(
                    page_content=row["text"],
                    metadata={
                        "segment": row["segment"],
                        "start": row["start"],
                        "end": row["end"],
                        "title": row["title"],
                        "url": row["url"],
                    },
                )
                data.append(document)
    return data


class Assistant:
    def __init__(self, embeddings, llm, prompt, top_k=10):
        self._embeddings = embeddings

        self._vectorstore = InMemoryVectorStore(embeddings)
        self._rag = (
            {"context": RunnablePassthrough(), "question": RunnablePassthrough()}
            | prompt
            | llm
            | StrOutputParser()
        )
        self.top_k = top_k

    def _format_docs(self, docs):
        return "\n\n".join(doc.page_content for doc in docs)

    def index_data(self, docs):
        self._vectorstore.add_documents(docs)

    def find_context(self, question):
        return self._vectorstore.similarity_search(question, k=self.top_k)

    def ask(self, question):
        context = self._format_docs(self.find_context(question))
        return self._rag.invoke({"context": context, "question": question})


embeddings = AzureOpenAIEmbeddings(
    azure_endpoint=ENDPOINT,
    api_key=API_KEY,
    model=EMBEDDINGS_MODEL,
)

llm = AzureChatOpenAI(
    api_key=API_KEY,
    azure_endpoint=ENDPOINT,
    api_version=API_VERSION,
    model=GENERATION_MODEL,
    temperature=0,
    seed=43,
)

prompt = ChatPromptTemplate.from_messages(
    messages=[
        (
            "system",
            "You are assistant to help user tu anser questions based on privided context. ",
        ),
        ("human", "context: '''{context}'''"),
        ("human", "{question}"),
    ]
)

documents_a = load_data("data/transcriptions")
documents_b = group_documents(documents_a, 5)
len(documents_a), len(documents_b)


(2585, 531)

In [5]:
%%time
###
### NOTE: RUN IT ONLY ONCE TO INDEX THE DATA (IT MAY TAKE UP TO 30 SECONDS)
###
rag_a = Assistant(embeddings, llm, prompt)
rag_a.index_data(documents_a)

rag_b = Assistant(embeddings, llm, prompt)
rag_b.index_data(documents_b)

CPU times: user 869 ms, sys: 162 ms, total: 1.03 s
Wall time: 11.7 s


## 2. Overview of the RAG System

You will have two RAG implementations, each with two key methods:

1.	find_context(query): Retrieves the most relevant transcript snippets for the query.
2.	ask(query): Generates an answer using the retrieved context.

Your job is to figure out which implementation (RAG `rag_a` vs. `rag_b`) performs better, using a well-defined set of evaluation criteria.


In [7]:
from pprint import pprint

QUESTION = "What is DIAL?"

print('RAG A (fine-grained chunks)')

print('\nAnswer (rag_a): \n')
pprint(rag_a.ask(QUESTION))

print('\nContext (rag_a): \n')
pprint(rag_a.find_context(QUESTION))

print()
print('RAG B (grouped chunks, n=5)')

print('\nAnswer (rag_b): \n')
pprint(rag_b.ask(QUESTION))

print('\nContext (rag_b): \n')
pprint(rag_b.find_context(QUESTION))

RAG A (fine-grained chunks)

Answer (rag_a): 

('Based on the provided context, **DIAL** appears to be an open-source AI '
 'framework or platform. It is described as having an "API-first approach" and '
 'is referred to as a "talk to your data framework," suggesting that it '
 'facilitates interaction with data, likely through conversational or '
 'AI-driven methods. Additionally, it mentions "Dial\'s implementation" and '
 '"Dial application itself," which implies that DIAL is both a framework and a '
 'tool for building applications.')

Context (rag_a): 

[Document(id='18884374-99a3-4e2c-a966-b4ff82c1afe1', metadata={'segment': 8, 'start': '00:00:25.750', 'end': '00:00:25.760', 'title': 'AI DIAL Overview', 'url': 'https://www.youtube.com/watch?v=2DlK5iMJLFk'}, page_content='applications dial stands for'),
 Document(id='9d83f6df-9903-4cd9-a93d-dad44a37dce6', metadata={'segment': 5, 'start': '00:00:15.470', 'end': '00:00:15.480', 'title': 'DIAL Quick Apps', 'url': 'https://www.youtube

## 3 Test Data Preparation

### 3.1	Examine the given video transcripts:
- Familiarize yourself with their [content](https://www.youtube.com/channel/UCmy-241n-ZLO4c5BA5W516g).
- Understand the format of the transcripts `data/transcriptions`.

### 3.2	Create a test dataset:
- Select or write a set of questions that you believe can be answered using the transcripts.
- Each question should have a ground-truth (expected) answer.
- Add relevand context to each question, which will be used to evaluate the RAG systems.
- Organize your dataset in a simple format: for each question, provide the corresponding ground-truth answer.


Example: 
```json
{
    "questions": [
        {
            "question": "What is DIAL?",
            "answer": "DIAL is ...",
            "relevant_context": [
                {
                    "transcript_id": "",
                    "start_time": "",
                    "end_time": ""
                }
            ]
        },
        {
            "question": "How to share DIAL chat with an other user?",
            "answer": "To share chat...",
            "relevant_context": [
                {
                    "transcript_id": "",
                    "start_time": "",
                    "end_time": ""
                }
            ]
        }, 
    ]
}
```

**Deliverable:** A JSON file (or even a Python dictionary) containing your test questions and answers.

In [8]:
import json

with open("test_dataset.json", "r", encoding="utf-8") as f:
    test_dataset = json.load(f)

print(json.dumps(test_dataset, indent=2))

{
  "questions": [
    {
      "question": "What is DIAL?",
      "answer": "DIAL stands for Deterministic Integrator of Applications and is an open-source AI framework with an API-first approach for building custom applications to tackle complex problems using LLMs.",
      "relevant_context": [
        {
          "transcript_id": "AI DIAL Overview",
          "start_time": "00:00:18.590",
          "end_time": "00:00:28.480"
        },
        {
          "transcript_id": "AI DIAL Overview",
          "start_time": "00:02:30.910",
          "end_time": "00:02:40.120"
        }
      ]
    },
    {
      "question": "What is the DIAL Marketplace?",
      "answer": "The DIAL Marketplace is a centralized hub of all accessible models, applications, and assistants within the DIAL enterprise ecosystem.",
      "relevant_context": [
        {
          "transcript_id": "DIAL Marketplace",
          "start_time": "00:00:08.950",
          "end_time": "00:00:18.880"
        }
      ]
    },


## 4. Select Evaluation Metrics

To measure the performance of each RAG implementation, you’ll use a combination of retrieval and generation metrics. Below are some examples, but feel free to choose others that you think are more appropriate.:

1.	Retrieval Metrics (evaluates find_context method):
- Recall@k: The proportion of relevant transcript snippets contained within the top-k retrieved snippets.
- Precision@k: How many of the retrieved snippets are actually relevant to the question.
- Mean Average Precision (MAP): The average precision across all questions.

2.	Generation Metrics (evaluates generate_answer method):
- ROUGE: Measures overlap between the generated answer and the ground-truth reference.
- BLEU: Measures how close the generated answer is to the ground-truth, based on overlapping n-grams.
- LLM-based metrics: Metrics that evaluate the fluency and coherence of the generated answer.

For a practical exercise, it’s common to pick Recall@k for retrieval and ROUGE or BLEU for generation.

**Deliverable:** A document or notebook section describing which metrics you have chosen and why.

**Hint:** Think also about some metrics to measure tone or style of the generated answers.

## Retrieval Metrics (find_context)

1. **Recall@k** — measures how many of the ground-truth relevant segments appear in the top-k
   retrieved documents. This is critical because if the context is missing, the answer will be wrong
   regardless of the LLM quality.

2. **Precision@k** — measures how many of the top-k retrieved documents are actually relevant.
   Avoids flooding the LLM with irrelevant context, which can confuse or distract the model.

### Generation Metrics (ask)

3. **ROUGE-L** — measures the longest common subsequence between the generated answer and the
   ground-truth. Good for checking if key facts and phrasing are captured without requiring exact word order.

4. **BLEU** — measures n-gram overlap between the generated and reference answers.
   Sensitive to exact wording, useful as a complementary signal to ROUGE.

5. **LLM-as-a-Judge (Faithfulness + Relevance)** — uses GPT to score whether the generated answer
   is faithful to the provided context and relevant to the question (0-1 scale). This catches
   hallucinations that lexical metrics like ROUGE/BLEU miss entirely.

### Why this combination?
- Recall@k tells us if retrieval can find what is needed.
- ROUGE-L + BLEU give fast, deterministic generation quality scores.
- LLM-as-a-Judge catches semantic quality that string-matching metrics miss.

## 5. Implement Metrics Calculation

1.	Implement or import the retrieval metrics:
- Evaluate how well find_context surfaces relevant transcript snippets.
- For each question, check how many relevant snippets appear in the top-k results.
- Aggregate the results across all questions.
	
2.	Implement or import the generation metrics:
- Compare the generated answer with the ground-truth by calculating selected metrics.

**Deliverable:** A Python code cells in your notebook where you implement or call functions to compute these metrics.

In [10]:
!pip install rouge-score nltk

Looking in indexes: https://pypi.org/simple, https://auto_epm-ease_git_integration%40epam.com:****@eu.artifactory.epam.com/artifactory/api/pypi/EPM-ESP-pypi/simple
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.8/135.8 kB 9.6 MB/s eta 0:00:00
  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24934 sha256=ac811942b708047452f32e7a43c1e74670f5a7d71aa83984b26ccee30e52a053
  Stored in directory: /home/jovyan/.cache/pip/wheels/1e/19/43/8a442dc83660ca25e163e1bd1f89919284ab0d0c1475475148
Successfully built rouge-score


In [11]:
import math
from rouge_score import rouge_scorer
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
import nltk
nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True)


# Retrieval Metrics

def is_relevant(doc: Document, relevant_context: list) -> bool:
    """Check if a retrieved document matches any ground-truth relevant context entry."""
    title = doc.metadata.get("title", "")
    start = doc.metadata.get("start", "")
    end   = doc.metadata.get("end", "")
    for rc in relevant_context:
        if rc["transcript_id"] == title:
            # Accept if time ranges overlap or the segment falls within the range
            if rc["start_time"] <= end and rc["end_time"] >= start:
                return True
    return False


def recall_at_k(retrieved_docs: list, relevant_context: list) -> float:
    """Recall@k: fraction of relevant contexts found in retrieved_docs."""
    if not relevant_context:
        return 1.0
    hits = sum(1 for rc in relevant_context
               if any(is_relevant(doc, [rc]) for doc in retrieved_docs))
    return hits / len(relevant_context)


def precision_at_k(retrieved_docs: list, relevant_context: list) -> float:
    """Precision@k: fraction of retrieved docs that are relevant."""
    if not retrieved_docs:
        return 0.0
    hits = sum(1 for doc in retrieved_docs if is_relevant(doc, relevant_context))
    return hits / len(retrieved_docs)


def average_precision(retrieved_docs: list, relevant_context: list) -> float:
    """Average Precision for a single query."""
    hits = 0
    precision_sum = 0.0
    for i, doc in enumerate(retrieved_docs, start=1):
        if is_relevant(doc, relevant_context):
            hits += 1
            precision_sum += hits / i
    total_relevant = len(relevant_context)
    return precision_sum / total_relevant if total_relevant > 0 else 0.0

# Generation Metrics

_rouge = rouge_scorer.RougeScorer(["rougeL"], use_stemmer=True)
_smooth = SmoothingFunction().method1


def rouge_l(generated: str, reference: str) -> float:
    """ROUGE-L F1 score."""
    return _rouge.score(reference, generated)["rougeL"].fmeasure


def bleu(generated: str, reference: str) -> float:
    """Sentence BLEU (1-4 gram) with smoothing."""
    ref_tokens = reference.lower().split()
    gen_tokens = generated.lower().split()
    return sentence_bleu([ref_tokens], gen_tokens, smoothing_function=_smooth)


def llm_judge(question: str, answer: str, context: str, llm) -> dict:
    """
    Use the LLM to rate faithfulness and relevance on a 0-1 scale.
    Returns {'faithfulness': float, 'relevance': float}
    """
    judge_prompt = f"""You are an evaluation judge. Score the following answer on two criteria.
Return ONLY a JSON object with two keys: "faithfulness" and "relevance", each a float between 0 and 1.

Question: {question}
Context: {context}
Answer: {answer}

Faithfulness: Is the answer fully supported by the context? (1=fully, 0=hallucinated)
Relevance: Does the answer directly address the question? (1=fully relevant, 0=irrelevant)

JSON:"""
    import json as _json
    response = llm.invoke(judge_prompt).content.strip()
    
    start = response.find("{")
    end   = response.rfind("}") + 1
    try:
        scores = _json.loads(response[start:end])
        return {"faithfulness": float(scores.get("faithfulness", 0)),
                "relevance":    float(scores.get("relevance", 0))}
    except Exception:
        return {"faithfulness": 0.0, "relevance": 0.0}

## 6. Compare Two RAG Implementations

### 6.1 Run the evaluation for rag_a and rag_b on your test dataset. Compare the results across all metrics.


### 6.2 Select the best performer:
- Provide a short analysis explaining why one might outperform the other.
- Highlight any trade-offs

**Deliverable:** Code and table comparing performance across metrics. A short report or notebook cell summarizing which RAG is “best” overall.



In [12]:
def evaluate_rag(rag, test_dataset: dict, llm, use_llm_judge: bool = True) -> pd.DataFrame:
    rows = []
    for item in test_dataset["questions"]:
        question        = item["question"]
        ground_truth    = item["answer"]
        relevant_ctx    = item["relevant_context"]

        retrieved_docs  = rag.find_context(question)
        generated_answer = rag.ask(question)
        context_text    = "\n\n".join(d.page_content for d in retrieved_docs)

        rec  = recall_at_k(retrieved_docs, relevant_ctx)
        prec = precision_at_k(retrieved_docs, relevant_ctx)
        ap   = average_precision(retrieved_docs, relevant_ctx)
        rl   = rouge_l(generated_answer, ground_truth)
        bl   = bleu(generated_answer, ground_truth)

        row = {
            "question":    question,
            "recall@k":    rec,
            "precision@k": prec,
            "avg_precision": ap,
            "rouge_l":     rl,
            "bleu":        bl,
        }

        if use_llm_judge:
            scores = llm_judge(question, generated_answer, context_text, llm)
            row["llm_faithfulness"] = scores["faithfulness"]
            row["llm_relevance"]    = scores["relevance"]

        rows.append(row)

    return pd.DataFrame(rows)

In [13]:
print("Evaluating RAG A (fine-grained chunks)...")
results_a = evaluate_rag(rag_a, test_dataset, llm, use_llm_judge=True)

print("Evaluating RAG B (grouped chunks, n=5)...")
results_b = evaluate_rag(rag_b, test_dataset, llm, use_llm_judge=True)

# per-question tables 
print("\n=== RAG A — per-question scores ===")
display(results_a.set_index("question"))

print("\n=== RAG B — per-question scores ===")
display(results_b.set_index("question"))

# aggregate comparison table 
metric_cols = ["recall@k", "precision@k", "avg_precision", "rouge_l", "bleu",
               "llm_faithfulness", "llm_relevance"]
summary = pd.DataFrame({
    "RAG A (fine-grained)": results_a[metric_cols].mean(),
    "RAG B (grouped n=5)":  results_b[metric_cols].mean(),
})
summary["delta (B - A)"] = summary["RAG B (grouped n=5)"] - summary["RAG A (fine-grained)"]
print("\n=== Aggregate comparison (mean across all questions) ===")
display(summary.round(4))

Evaluating RAG A (fine-grained chunks)...
Evaluating RAG B (grouped chunks, n=5)...

=== RAG A — per-question scores ===


,recall@k,precision@k,avg_precision,rouge_l,bleu,llm_faithfulness,llm_relevance
question,,,,,,,
What is DIAL?,0.5,0.1,0.500000,0.228571,0.043450,0.7,0.9
What is the DIAL Marketplace?,1.0,0.1,0.333333,0.168675,0.017690,0.6,0.8
What programming framework is used to develop DIAL applications?,0.5,0.1,0.071429,0.200000,0.030935,0.5,0.8
How can DIAL be integrated with Microsoft Copilot Studio?,0.0,0.0,0.000000,0.074468,0.020070,0.5,0.8
What are DIAL Quick Apps?,1.0,0.1,0.166667,0.271605,0.060841,0.7,0.9
How can DIAL be deployed from Google Cloud Marketplace?,0.0,0.0,0.000000,0.296296,0.082626,0.5,0.7



=== RAG B — per-question scores ===


,recall@k,precision@k,avg_precision,rouge_l,bleu,llm_faithfulness,llm_relevance
question,,,,,,,
What is DIAL?,1.0,0.2,0.700000,0.260870,0.020699,1.0,1.0
What is the DIAL Marketplace?,1.0,0.1,1.000000,0.243902,0.091149,0.9,1.0
What programming framework is used to develop DIAL applications?,0.0,0.0,0.000000,0.222222,0.033744,1.0,1.0
How can DIAL be integrated with Microsoft Copilot Studio?,0.0,0.0,0.000000,0.087805,0.007410,0.9,1.0
What are DIAL Quick Apps?,1.0,0.1,0.333333,0.197674,0.072800,0.9,1.0
How can DIAL be deployed from Google Cloud Marketplace?,0.0,0.0,0.000000,0.149068,0.036499,0.9,1.0



=== Aggregate comparison (mean across all questions) ===


,RAG A (fine-grained),RAG B (grouped n=5),delta (B - A)
recall@k,0.5000,0.5000,0.0000
precision@k,0.0667,0.0667,0.0000
avg_precision,0.1786,0.3389,0.1603
rouge_l,0.2066,0.1936,-0.0130
bleu,0.0426,0.0437,0.0011
llm_faithfulness,0.5833,0.9333,0.3500
llm_relevance,0.8167,1.0000,0.1833


### Metric-by-Metric Verdict- RAG B (grouped chunks, n=5)

**RAG B wins on the metrics that matter most in production:**
- **MAP (+0.16):** RAG B ranks relevant documents higher within the top-k, meaning the LLM gets the most important context earlier.
- **LLM Faithfulness (+0.35):** The biggest gap — RAG B answers are almost
  never hallucinated (0.93 vs 0.58)
- **LLM Relevance (+0.18):** RAG B answers are fully on-topic for every question (1.0).

RAG B creates **richer, self-contained context chunks**.
The LLM receives complete paragraphs instead of isolated one-sentence fragments,which dramatically reduces hallucination and improves answer coherence.
RAG A's fine-grained chunks sometimes provide incomplete context(0.5 faithfulness on "How can DIAL be integrated with Microsoft Copilot Studio?"),forcing the LLM to fill gaps from its parametric knowledge.

## Step 7: Propose Improvements

Based on your analysis, suggest ways to improve evaluation metrics or the RAG systems themselves. For example, you might propose a new metric, a new model architecture, or a different way to combine retrieval and generation. Additionally think how to deal with randomness in the results.

**Deliverable:** A short report or notebook cell with your suggestions for improvement.

### 1. Handle Randomness in Results
- `temperature=0` + `seed=43` already set — but LLM-Judge scores still varied
  (RAG A faithfulness ranged 0.5–0.7 per question).
- Average multiple judge calls per question to reduce LLM-Judge noise.

### 2. Fix Recall=0.0 on 2 Questions 
Both RAG A and RAG B scored **0.0 recall** on:
  - "How can DIAL be integrated with Microsoft Copilot Studio?"
  - "How can DIAL be deployed from Google Cloud Marketplace?"
These are proper nouns — dense embeddings fail here.
- **Hybrid search** (BM25 + dense): BM25 keyword match would surface exact names.
- **Increase k** from 10 to 20 and plot Recall@k curve.
- **Sliding-window chunking** (size=5, step=2): overlapping chunks prevent
  relevant sentences from being buried.

### 3. Improve Faithfulness 
RAG A's short fragments force the LLM to fill gaps from parametric memory.
- Add prompt instruction: "Answer ONLY using the provided context.
  If the answer is not in the context, say I don't know."
  
### 4. Better Metrics to Replace Low BLEU Scores
BLEU scores were 0.02–0.09 for both systems — misleadingly low because
correct paraphrased answers are penalized.
- **BERTScore**: semantic similarity via embeddings — handles paraphrasing correctly.
- **RAGAS framework**: structured pipeline with context precision, context recall,
  answer faithfulness, answer relevance — more granular than current metrics.
- **Hallucination rate**: binary flag (faithfulness < 0.5) —
  RAG A fails 3/6 questions, RAG B fails 0/6 by this threshold.
